# TRABAJO INTEGRADOR


Curso: Introducción a Análisis de Datos

Comisión: C44

Alumna: Sofia Lara Goszko


## Instalación e importación de librerías

Comenzamos instalando e importando las librerías que vamos a necesitar.

In [1]:
!pip install pandas==2.2.2

In [2]:
!pip install numpy

In [3]:
import pandas as pd #importamos la librería pandas y le colocamos el alias "pd"
import numpy as np #importamos la librería numpy y le colocamos el alias "np"

## Importación de datasets originales

Importamos los datasets `.csv` que vamos a utilizar para realizar el análisis.

### CUSTOMERS

Importamos el dataset `ecommerce_customers_dataset.csv`

In [4]:
df_clientes = pd.read_csv("./datasets_originales/ecommerce_customers_dataset.csv") #importamos el dataset
df_clientes.head() #imprimimos sus primeros 5 renglones para ver las columnas del dataset

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

### ORDER ITEMS

Importamos el dataset `ecommerce_order_items_dataset.csv`

In [5]:
df_order_items = pd.read_csv("./datasets_originales/ecommerce_order_items_dataset.csv") #importamos el dataset
df_order_items.head() #imprimimos sus primeros 5 renglones para ver las columnas del dataset

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

### ORDER PAYMENTS
 
Importamos el dataset `ecommerce_order_payments_dataset.csv`

In [6]:
df_order_payments = pd.read_csv("./datasets_originales/ecommerce_order_payments_dataset.csv") #importamos el dataset
df_order_payments.head() #imprimimos sus primeros 5 renglones para ver las columnas del dataset

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

### ORDERS

Importamos el dataset `ecommerce_orders_dataset.csv`

In [7]:
df_orders = pd.read_csv("./datasets_originales/ecommerce_orders_dataset.csv") #importamos el dataset
df_orders.head() #imprimimos sus primeros 5 renglones para ver las columnas del dataset

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

### PRODUCTS 

Importamos el dataset `ecommerce_products_dataset.csv`

In [8]:
df_productos = pd.read_csv("./datasets_originales/ecommerce_products_dataset.csv") #importamos el dataset
df_productos.head() #imprimimos sus primeros 5 renglones para ver las columnas del dataset

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

## Datos de los datasets

Una vez importados, analizamos cómo son cada uno de los dataset (shape, tipo de datos y cantidad de nulos).

In [9]:
def mostrarDatos(df):
    print(f"Tamaño del dataset: {df.shape}") #permite ver la forma de dataset
    print(f"---------")
    print(f"---------")
    print(f"Información del dataset")
    print(df.info()) #permite la cantidad de no nulos y el tipo de dato de cada columna del dataset
    print(f"---------")
    print(f"---------")
    print(f"Recuento de nulos") #permite ver la cantidad de nulos que hay en cada dataset
    print(df.isnull().sum())

### CUSTOMERS

Veamos cómo está el `df_clientes`



In [10]:
mostrarDatos(df_clientes)

Tamaño del dataset: (99441, 5)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
---------
---------
Recuento de nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64


El dataset de clientes no posee nulos pero si habrá que modificar los tipos de dato de algunas columnas para su posterior analisis.

### ORDER ITEMS

Veamos cómo está el `df_order_items`


In [11]:
mostrarDatos(df_order_items)

Tamaño del dataset: (112650, 7)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
---------
---------
Recuento de nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64


El dataset de clientes no posee nulos pero si habrá que modificar los tipos de dato de algunas columnas para su posterior analisis.

### ORDER PAYMENTS

Veamos cómo está el `df_order_payments`


In [12]:
mostrarDatos(df_order_payments)

Tamaño del dataset: (103886, 5)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
---------
---------
Recuento de nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


El dataset de clientes no posee nulos pero si habrá que modificar los tipos de dato de algunas columnas para su posterior analisis.

### ORDERS

Veamos cómo está el `df_orders`



In [13]:
mostrarDatos(df_orders)

Tamaño del dataset: (99441, 8)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
---------
---------
Recuento de nulos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp      

En ese caso el dataset si tiene nulos, en la próxima sección se verá su tratamiento además de modificar el tipo de dato de las columnas para su posterior análisis.

### PRODUCTS

Veamos cómo está el `df_productos`

In [14]:
mostrarDatos(df_productos)

Tamaño del dataset: (32951, 9)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
---------
---------
Recuento de nulos
product_id                      0
product_category_name         610
product_name_lenght        

Sucede lo mismo que en el dataset anterior

## Limpieza de datos

En esta sección se limpiarán los datasets:

- Tratamiento de nulo

- Cambio de tipo de datos

Para el primero, se analizará el porcentaje de nulos en cada columna. Si no supera el 5% se eliminará la fila.



In [15]:
def columnasSinNulos(df):
    porcentaje_nulos = (df.isnull().sum() / len(df)) * 100 #calcula el % de nulos de cada columna del df
    print(porcentaje_nulos)
    columnas_sin_nulos = porcentaje_nulos[porcentaje_nulos <= 5].index #se queda solo con las columnas con un % de nulos <= al 5
    return columnas_sin_nulos #devuelve las columnas filtradas

def eliminarNulos(df, columnas):
    df= df.dropna(subset=columnas, inplace=True) #el df sin nulos reemplaza al anterior
    return df

In [16]:
#transforma a tipo fecha
def transformarAFecha(df, columnas): #parametros (DataFrame, columnas a modificar)
    df[columnas] = df[columnas].apply(pd.to_datetime) #convierto al tipo de dato fecha
    return df    

In [17]:
#transforma a tipo string
def transformarAString(df, columnas): #parametros (DataFrame, columnas a modificar)
    for columna in columnas: #itero sobre las columnas
        df[columna] = df[columna].astype("string") #convierto al tipo de dato string
    return df    

In [18]:
#transforma todo a mayursculas
def transformarAMayuscula(df, columnas): #parametros (DataFrame, columnas a modificar)
    for columna in columnas: #itero sobre las columnas
        df[columna] = df[columna].str.upper() #convierto todo a mayusculas
    return df    

### CUSTOMERS

In [33]:
columnas = ["customer_id", "customer_unique_id", "customer_zip_code_prefix", "customer_city", "customer_state"] #columnas a convertir a string
transformarAString(df_clientes, columnas)
columnas = ["customer_city", "customer_state"] #columnas a pasar a mayusculas
transformarAMayuscula(df_clientes, columnas)

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

      customer_zip_code_prefix          customer_city customer_state  
0                        14409                 FRANCA             SP  
1                         9790  SAO BERNARDO DO CAMPO             SP  
2                         1151              SAO PAULO             SP  
3                         8775        MOGI DAS CRUZES             SP  
4                        13056               CAMPINAS             SP  
...                        ...                    ...            ...  
99436                     3937              SAO PAULO             SP  
99437                     6764        TABOAO DA SERRA             SP  
99438                    60115              FORTALEZA             CE  
99439                    92120                 CANOAS             RS  
99440                     6703                  COTIA             SP  

[99441 rows x 5 columns]

In [34]:
#reviso que los cambios se hayan realizado 
mostrarDatos(df_clientes)

Tamaño del dataset: (99441, 5)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB
None
---------
---------
Recuento de nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64


### ORDER ITEMS

In [21]:

columnas = ["order_id", "order_item_id", "product_id", "seller_id"] #columnas a convetir a string
transformarAString(df_order_items, columnas)
columnas = ["shipping_limit_date"]
transformarAFecha(df_order_items, columnas) #columna a convertir a date

order_id order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214             1   
1       00018f77f2f0320c557190d7a144bdd3             1   
2       000229ec398224ef6ca0657da4fc703e             1   
3       00024acbcdf0a6daa1e931b038114c75             1   
4       00042b26cf59d7ce69dfabb4e55b4fd9             1   
...                                  ...           ...   
112645  fffc94f6ce00a00581880bf54a75a037             1   
112646  fffcd46ef2263f404302a634eb57f7eb             1   
112647  fffce4705a9662cd70adb13d4a31832d             1   
112648  fffe18544ffabc95dfada21779c9644f             1   
112649  fffe41c64501cc87c801fd61db3f6244             1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

       shipping_limit_date   price  freight_value  
0      2017-09-19 09:45:35   58.90          13.29  
1      2017-05-03 11:05:13  239.90          19.93  
2      2018-01-18 14:48:30  199.00          17.87  
3      2018-08-15 10:10:18   12.99          12.79  
4      2017-02-13 13:57:51  199.90          18.14  
...                    ...     ...            ...  
112645 2018-05-02 04:11:01  299.99          43.41  
112646 2018-07-20 04:31:48  350.00          36.53  
112647 2017-10-30 17:14:25   99.90          16.95  
112648 2017-08-21 00:04:32   55.99           8.72  
112649 2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

In [22]:
#reviso que los cambios se hayan realizado 
mostrarDatos(df_order_items)

Tamaño del dataset: (112650, 7)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  string        
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), string(4)
memory usage: 6.0 MB
None
---------
---------
Recuento de nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtyp

### ORDER PAYMENTS

In [23]:
columnas = ["order_id", "payment_type"] #columnas a convertir en string
transformarAString(df_order_payments, columnas) 
transformarAMayuscula(df_order_payments, ["payment_type"]) #columna a convetrir en mayuscula

order_id  payment_sequential payment_type  \
0       b81ef226f3fe1789b1e8b2acac839d17                   1  CREDIT_CARD   
1       a9810da82917af2d9aefd1278f1dcfa0                   1  CREDIT_CARD   
2       25e8ea4e93396b6fa0d3dd708e76c1bd                   1  CREDIT_CARD   
3       ba78997921bbcdc1373bb41e913ab953                   1  CREDIT_CARD   
4       42fdf880ba16b47b59251dd489d4441a                   1  CREDIT_CARD   
...                                  ...                 ...          ...   
103881  0406037ad97740d563a178ecc7a2075c                   1       BOLETO   
103882  7b905861d7c825891d6347454ea7863f                   1  CREDIT_CARD   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  CREDIT_CARD   
103884  b8b61059626efa996a60be9bb9320e10                   1  CREDIT_CARD   
103885  28bbae6599b09d39ca406b747b6632b1                   1       BOLETO   

        payment_installments  payment_value  
0                          8          99.33  
1                          1          24.39  
2                          1          65.71  
3                          8         107.78  
4                          2         128.45  
...                      ...            ...  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58  

[103886 rows x 5 columns]

In [24]:
#reviso que los datos se hayan realizado
mostrarDatos(df_order_payments)

Tamaño del dataset: (103886, 5)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB
None
---------
---------
Recuento de nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


### ORDERS

In [25]:
#tratamiento de nulos
columnas = columnasSinNulos(df_orders) #devuelve las columnas filtradas
eliminarNulos(df_orders, columnas) #elimina los nulos

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64


In [26]:
columnas = ["order_id", "customer_id", "order_status"] #columnas a transformar a string
transformarAString(df_orders, columnas)
transformarAMayuscula(df_orders, ["order_status"]) #columna a convertir a mayusculas
columnas = ["order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date", "order_delivered_customer_date", "order_estimated_delivery_date"] #columnas a converitr a datetime
transformarAFecha(df_orders, columnas)

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        DELIVERED      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        DELIVERED      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        DELIVERED      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        DELIVERED      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        DELIVERED      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    DELIVERED      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    DELIVERED      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    DELIVERED      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    DELIVERED      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    DELIVERED      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0                        2017-10-18  
1                        2018-08-13  
2                        2018-09-04  
3                        2017-12-15  
4                        2018-02-26  
...                             ...  
99436                    2017-03-28  
99437                    2018-03-02  
99438                    2017-09-27  
99439                    2018-02-15  
99440                    2018-04-03  

[96461 rows x 8 columns]

In [27]:
#reviso que se hayan hecho los cambios
mostrarDatos(df_orders)

Tamaño del dataset: (96461, 8)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB
None
---------
---------
Recuento de nulos
order_id                         0
customer_id     

### PRODUCTS

In [28]:
#tratamiento de nulos
columnas = columnasSinNulos(df_productos) #devuelve las columnas filtradas
eliminarNulos(df_productos, columnas) #elimina los nulos

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64


In [29]:
columnas = ["product_id", "product_category_name"] #columnas a transformar a string
transformarAString(df_productos, columnas)
transformarAMayuscula(df_productos, ["product_category_name"]) #columna a poner en mayuscula

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         PERFUMARIA   
1      3aa071139cb16b67ca9e5dea641aaa2f                              ARTES   
2      96bd76ec8810374ed1b65e291975717f                      ESPORTE_LAZER   
3      cef67bcfe19066a932b7673e239eb23d                              BEBES   
4      9dc1a7de274444849c219cff195d0b71              UTILIDADES_DOMESTICAS   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   MOVEIS_DECORACAO   
32947  bf4538d88321d0fd4412a93c974510e6  CONSTRUCAO_FERRAMENTAS_ILUMINACAO   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    CAMA_MESA_BANHO   
32949  83808703fc0706a22e264b9d75f04a2e             INFORMATICA_ACESSORIOS   
32950  106392145fca363410d287a815be6de4                    CAMA_MESA_BANHO   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32340 rows x 9 columns]

In [30]:
#reviso que se hayan hecho los cambios
mostrarDatos(df_productos)

Tamaño del dataset: (32340, 9)
---------
---------
Información del dataset
<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB
None
---------
---------
Recuento de nulos
product_id                    0
product_category_name         0
product_name_lenght           0
produ

## Analisis de datos

Una vez limpiados, se procede a analizar los datos

In [73]:
def analisisDatos(df): 
  print("----------")
  print("Análisis cuantitativo")
  columnas_numericas = df.select_dtypes(include=["int", "float"]).columns #se queda con las columnas numericas
  if not columnas_numericas.empty: #si no hay columnas vacias
    print(df[columnas_numericas].describe()) #se usa el decribe en cada columna
  print("----------")
  print("----------")
  print("Análisis variables cualitativas")
  columnas_string = df.select_dtypes(include=["string"]).columns #se queda con las variables tipo string y datetime
  if not columnas_string.empty: #si no hay columnas vacias
    print(df[columnas_string].describe()) #se usa el decribe en cada columa

### CUSTOMERS

In [74]:
analisisDatos(df_clientes)

----------
Análisis cuantitativo
----------
----------
Análisis variables cualitativas
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     SAO PAULO             SP  
freq                        142         15540          41746  


Se puede observar que el estado de São Paulo, en especial la ciudad de São Paulo, es el mayor comprador. Hagamos una lupa en eso.

In [36]:
#veamos qué estados son los mayores compradores
df_clientes["customer_state"].value_counts()
     

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: Int64

Podemos ver que los clientes se encuentran en 3 estados principalemente: São Paulo, Rio de Janeiro y Minas Gerais.

In [44]:
#Analizando São Paulo
df_SP = df_clientes[df_clientes.customer_state=="SP"] #creamos un DataFrame solo de ese estado
df_SP["customer_city"].value_counts() #analizamos en que ciudades estan los principales compradores



customer_city
SAO PAULO                15540
CAMPINAS                  1444
GUARULHOS                 1189
SAO BERNARDO DO CAMPO      938
SANTO ANDRE                796
                         ...  
TERMAS DE IBIRA              1
AMERICO DE CAMPOS            1
PEREIRAS                     1
CASSIA DOS COQUEIROS         1
NATIVIDADE DA SERRA          1
Name: count, Length: 629, dtype: Int64

In [54]:
#veamos que porcentaje implica la ciudad de SAO PAULO sobre el total del estado
porcentaje = (len(df_SP[df_SP.customer_city=="SAO PAULO"]) / len(df_SP)) *100
print(porcentaje)

37.22512336511283


In [57]:
#Analizando Rio de Janeiro
df_RJ = df_clientes[df_clientes.customer_state=="RJ"] #creamos un DataFrame solo de ese estado
df_RJ["customer_city"].value_counts() #analizamos en que ciudades estan los principales compradores


customer_city
RIO DE JANEIRO               6882
NITEROI                       849
NOVA IGUACU                   442
SAO GONCALO                   409
DUQUE DE CAXIAS               266
                             ... 
CORREGO DO OURO                 1
ITACURUSSA                      1
SANTA ISABEL DO RIO PRETO       1
BOA ESPERANCA                   1
SAO SEBASTIAO DO PARAIBA        1
Name: count, Length: 149, dtype: Int64

In [60]:
#veamos que porcentaje implica la ciudad de RIO DE JANEIRO sobre el total del estado
porcentaje = (len(df_RJ[df_RJ.customer_city=="RIO DE JANEIRO"]) / len(df_RJ)) *100
print(porcentaje)

53.54808590102708


In [61]:
#Analizando Mina Gerais
df_MG = df_clientes[df_clientes.customer_state=="MG"] #creamos un DataFrame solo de ese estado
df_MG["customer_city"].value_counts() #analizamos en que ciudades estan los principales compradores

customer_city
BELO HORIZONTE             2773
JUIZ DE FORA                427
CONTAGEM                    426
UBERLANDIA                  374
MONTES CLAROS               211
                           ... 
CASCALHO RICO                 1
CONFINS                       1
SAO FRANCISCO DO GLORIA       1
RIO ESPERA                    1
SERRA BONITA                  1
Name: count, Length: 745, dtype: Int64

In [101]:
#veamos que porcentaje implica la ciudad de BELO HORIZONTE sobre el total del estado
porcentaje = (len(df_MG[df_MG.customer_city=="BELO HORIZONTE"]) / len(df_MG)) *100
print(porcentaje)

23.83326171035668


Como se puede observar, los principales clientes se encuentran en las capitales de São Paulo (São Paulo), Rio de Janeiro (Rio de Janeiro) y Minas Gerais (Belo Horizonte).


### ORDER ITEMS

In [76]:
analisisDatos(df_order_items)

----------
Análisis cuantitativo
               price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000
----------
----------
Análisis variables cualitativas
                                order_id order_item_id  \
count                             112650        112650   
unique                             98666            21   
top     8272b63d03f5f79c56e9e4120aec44ef             1   
freq                                  21         98666   

                              product_id                         seller_id  
count                             112650                            112650  
unique                             32951                              3095  
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44

Del análisis cuantitativo se puede observar que, en su mayoría, se realizan compras menores.

En promedio los clientes gastan 120 reales y sólo el 25% paga mas de 135 reales.

### ORDER PAYMENTES

In [84]:
analisisDatos(df_order_payments)

----------
Análisis cuantitativo
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
----------
----------
Análisis variables cualitativas
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  CREDIT_CARD
freq                                  29        76795


In [85]:
#analizando los tipos de pago
df_order_payments["payment_type"].value_counts()


payment_type
CREDIT_CARD    76795
BOLETO         19784
VOUCHER         5775
DEBIT_CARD      1529
NOT_DEFINED        3
Name: count, dtype: Int64

In [86]:
porcentaje = (len(df_order_payments[df_order_payments.payment_type=="CREDIT_CARD"])/len(df_order_payments))*100
print(porcentaje)

73.92237645110987


El método de pago más elegido es la tarjeta de crédito con casi el 74% del total.

### ORDERS

In [87]:
analisisDatos(df_orders)

----------
Análisis cuantitativo
----------
----------
Análisis variables cualitativas
                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       DELIVERED  
freq          96455  


In [88]:
#analizando el order_status
df_orders["order_status"].value_counts()

order_status
DELIVERED    96455
CANCELED         6
Name: count, dtype: Int64

In [90]:
porcentaje = (len(df_orders[df_orders.order_status=="DELIVERED"])/len(df_orders))*100
print(porcentaje)

99.9937798695846


Si bien el 99,99% de las ordenes se entregan antes de ser canceladas, no tiene una clientela regular.

### PRODUCTS

In [104]:
analisisDatos(df_productos)

----------
Análisis cuantitativo
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.636115   
min 

In [105]:
#veamos las categorias de los productos
df_productos["product_category_name"].value_counts()

product_category_name
CAMA_MESA_BANHO                  3029
ESPORTE_LAZER                    2867
MOVEIS_DECORACAO                 2657
BELEZA_SAUDE                     2444
UTILIDADES_DOMESTICAS            2335
                                 ... 
FASHION_ROUPA_INFANTO_JUVENIL       5
CASA_CONFORTO_2                     5
PC_GAMER                            3
SEGUROS_E_SERVICOS                  2
CDS_DVDS_MUSICAIS                   1
Name: count, Length: 73, dtype: Int64

Tomando en cuenta el análisis cuantitativo se puede observar que el 75% de los productos son pequeños: 38cm de largo, 21cm de alto y 30cm de ancho.

Asimismo, los vendedores suben suben en promedio 2 fotos por producto.

Del análisis cualitativo se obtiene que la categoría más elegida por los cliente fue la de cama_mesa_baño, seguida por deporte_ocio y muebles_decoración.

